vertical-pod-autoscaler/pkg/recommender/routines/safety_margin_modifier_post_processor.go

```diff
@@ -0,0 +1,221 @@
+				continue
+			}
+			newRec = logModifier(recCopy, r.DefaultSafetyMarginFactor, resourceModifier.Parameters[0])
+		case exponential:
+			if len(resourceModifier.Parameters) != 2 {
+				klog.Errorf("Skipping %s safety margin modifier in vpa %s/%s: exponential modifier requires 2 parameters, %d given", resourceName, name, vpa.ID.Namespace, vpa.ID.VpaName, len(resourceModifier.Parameters))
+				continue
+			}
+			newRec = exponentialModifier(recCopy, r.DefaultSafetyMarginFactor, resourceModifier.Parameters[0], resourceModifier.Parameters[1])
+		case undefined:
+			continue
+		default:
+			klog.Errorf("Skipping %s safety margin modifier in vpa %s/%s: specified modifier is not valid", name, vpa.ID.Namespace, vpa.ID.VpaName)
+		}
+
+		if resourceName == v1.ResourceCPU {
+			containerRec[resourceName] = *resource.NewMilliQuantity(int64(newRec), resource.DecimalSI)
+		} else {
+			containerRec[resourceName] = *resource.NewQuantity(int64(newRec), resource.DecimalSI)
+		}
+	}
+}
+
+// Undo default safety margin before applying a custom linear safety margin (a * baseRec)
+func linearModifier(rec resource.Quantity, defaultSafetyMargin, slope float64) float64 {
+	return float64(rec.Value()) / defaultSafetyMargin * slope
+}
+
+// Undo default safety margin before applying a custom affine safety margin (a * baseRec + b)
+func affineModifier(rec resource.Quantity, defaultSafetyMargin, constant, slope float64) float64 {
+	return float64(rec.Value())/defaultSafetyMargin*slope + constant
+}
+
+// Undo default safety margin before applying a custom logarithmic safety margin (baseRec + a * log10(baseRec))
+func logModifier(rec resource.Quantity, defaultSafetyMargin, factor float64) float64 {
+	rawRec := float64(rec.Value()) / defaultSafetyMargin
+	return rawRec + math.Log10(rawRec)*factor
+}
+
+// Undo default safety margin before applying a custom exponential safety margin (baseRec + a * baseRec^b)
+func exponentialModifier(rec resource.Quantity, defaultSafetyMargin, exponent, factor float64) float64 {
+	rawRec := float64(rec.Value()) / defaultSafetyMargin
+	return rawRec + math.Pow(rawRec, exponent)*factor
+}
+
+func readSafetyMarginModifierFromVPAAnnotations(vpa *model.Vpa) map[string]safetyMarginModifier {
+	modifiers := map[string]safetyMarginModifier{}
+	for key, value := range vpa.Annotations {
+		containerName := extractContainerName(key, vpaPostProcessorPrefix, vpaPostProcessorSafetyMarginModifierSuffix)
+		if containerName == "" {
+			continue
+		}
+
+		safetyMarginModifier := safetyMarginModifier{}
+		safetyMarginFunction := safetyMarginFunction{}
+		if err := json.Unmarshal([]byte(value), &safetyMarginModifier); err != nil {
+			if err := json.Unmarshal([]byte(value), &safetyMarginFunction); err != nil {
+				klog.Errorf("Skipping safety margin modifier definition '%s' for container '%s' in vpa %s/%s due to bad format, error:%#v", value, containerName, vpa.ID.Namespace, vpa.ID.VpaName, err)
```

Comment:
The user experience is not going to be great in case of syntax error.
I know that for the moment we don't have any support for events in that code, I am wondering how we should surface these errors to the user...

Ok to continue with that code for that PR, but we should think about improving UX in case of err.